pkg/yurthub/filter/nodeportisolation/filter.go

```diff
@@ -0,0 +1,238 @@
+
+	nodeKey, err := s.KeyFunc(storage.KeyBuildInfo{
+		Component: "kubelet",
+		Name:      nif.nodeName,
+		Resources: "nodes",
+		Group:     "",
+		Version:   "v1",
+	})
+	if err != nil {
+		return fmt.Errorf("failed to get node key for %s, %v", nif.nodeName, err)
+	}
+	nif.nodeSynced = func() bool {
+		obj, err := s.Get(nodeKey)
+		if err != nil || obj == nil {
+			return false
+		}
+
+		if _, ok := obj.(*v1.Node); !ok {
+			return false
+		}
+
+		return true
+	}
+
+	nif.nodeGetter = func(name string) (*v1.Node, error) {
+		key, err := s.KeyFunc(storage.KeyBuildInfo{
+			Component: "kubelet",
+			Name:      name,
+			Resources: "nodes",
+			Group:     "",
+			Version:   "v1",
+		})
+		if err != nil {
+			return nil, fmt.Errorf("nodeGetter failed to get node key for %s, %v", name, err)
+		}
+		obj, err := s.Get(key)
+		if err != nil {
+			return nil, err
+		} else if obj == nil {
+			return nil, fmt.Errorf("node(%s) is not ready", name)
+		}
+
+		if node, ok := obj.(*v1.Node); ok {
+			return node, nil
+		}
+
+		return nil, fmt.Errorf("node(%s) is not found", name)
+	}
+
+	return nil
+}
+
+func (nif *nodePortIsolationFilter) Filter(obj runtime.Object, stopCh <-chan struct{}) runtime.Object {
+	if ok := cache.WaitForCacheSync(stopCh, nif.nodeSynced); !ok {
+		return obj
+	}
+
+	switch v := obj.(type) {
+	case *v1.ServiceList:
+		var svcNew []v1.Service
+		for i := range v.Items {
+			svc := nif.isolateNodePortService(&v.Items[i])
+			if svc != nil {
+				svcNew = append(svcNew, *svc)
+			}
+		}
+		v.Items = svcNew
+		return v
+	case *v1.Service:
+		return nif.isolateNodePortService(v)
+	default:
+		return v
+	}
+}
+
+func (nif *nodePortIsolationFilter) isolateNodePortService(svc *v1.Service) *v1.Service {
+	nodePoolName := nif.nodePoolName
+	if len(nodePoolName) == 0 {
+		node, err := nif.nodeGetter(nif.nodeName)
+		if err != nil {
+			klog.Warningf("skip isolateNodePortService filter, failed to get node(%s), %v", nif.nodeName, err)
+			return svc
+		}
+		nodePoolName = node.Labels[nodepoolv1alpha1.LabelCurrentNodePool]
+	}
+
+	// node is not located in NodePool, skip this filter
+	if len(nodePoolName) == 0 {
+		return svc
+	}
+
+	nsName := fmt.Sprintf("%s/%s", svc.Namespace, svc.Name)
+	if svc.Spec.Type == v1.ServiceTypeNodePort {
+		if len(svc.Annotations[ServiceAnnotationNodePortIsolation]) != 0 {
+			enabledNodePools, disabledNodePools := parseNodePoolFromAnnotation(svc.Annotations[ServiceAnnotationNodePortIsolation])
+			if disabledNodePools.Has(nodePoolName) {
+				klog.V(2).Infof("nodePort service(%s) is isolated in nodePool(%s) by nodePortIsolationFilter", nsName, nodePoolName)
+				return nil
+			}
+
+			if enabledNodePools.Has(nodePoolName) {
+				return svc
+			}
+
+			if enabledNodePools.Has("*") {
+				return svc
+			}
+
+			klog.V(2).Infof("nodePort service(%s) is isolated in nodePool(%s) by nodePortIsolationFilter", nsName, nodePoolName)
+			return nil
```

Comment:
In other case, we return nil to disable the nodeport svc. It looks confusing. For example, we have "np1,-np2" which also implicitly disable this svc in np3, np4 and so on. Users may be confused when encountering this scenario.

Reply:
fixed